File: src/main/java/com/qtech/bubbles/data/SaveData.java
package com.qtech.bubbles.data;

import com.qtech.utilities.python.builtins.ValueError;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * @author Quinten Jungblut
 */
public class SaveData {
    public final HashMap<List<String>, Object> saveData = new HashMap<>();

    public String getString(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof String) {
            return (String) rawData;
        } else {
            throw new ValueError("Data is not instance of String");
        }
    }

    public Integer getInt(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof Integer) {
            return (Integer) rawData;
        } else {
            throw new ValueError("Data is not instance of Integer");
        }
    }

    public Float getFloat(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof Float) {
            return (Float) rawData;
        } else {
            throw new ValueError("Data is not instance of Float");
        }
    }

    public Double getDouble(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof Double) {
            return (Double) rawData;
        } else {
            throw new ValueError("Data is not instance of Double");
        }
    }

    public Boolean getBoolean(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof Boolean) {
            return (Boolean) rawData;
        } else {
            throw new ValueError("Data is not instance of Boolean");
        }
    }

    public Object getObject(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData != null) {
            return rawData;
        } else {
            throw new ValueError("Data is not instance of Object");
        }
    }

    @SuppressWarnings("unchecked")
    public <T> ArrayList<? extends T> getArrayList(String... path) {
        ArrayList<String> stringArray = new ArrayList<>(Arrays.asList(path));

        if (!saveData.containsKey(Arrays.asList(path))) {
            return null;
        }

        Object rawData = saveData.get(stringArray);
        if (rawData instanceof ArrayList) {
            return (ArrayList<T>) rawData;
        } else {
            throw new ValueError("Data is not instance of ArrayList");
        }
    }
}

File: src/main/kotlin/qtech/bubbles/data/SaveData.kt
package qtech.bubbles.data

import qtech.utilities.python.builtins.ValueError

/**
 * @author Quinten Jungblut
 */
@Suppress("unused")
class SaveData {
    val saveData = HashMap<List<String>, Any>()
    fun getString(vararg path: String): String? {
        val stringArray: ArrayList<String> = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is String) {
            rawData
        } else {
            throw ValueError("Data is not instance of String")
        }
    }

    fun getInt(vararg path: String): Int? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is Int) {
            rawData
        } else {
            throw ValueError("Data is not instance of Integer")
        }
    }

    fun getFloat(vararg path: String): Float? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is Float) {
            rawData
        } else {
            throw ValueError("Data is not instance of Float")
        }
    }

    fun getDouble(vararg path: String): Double? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is Double) {
            rawData
        } else {
            throw ValueError("Data is not instance of Double")
        }
    }

    fun getBoolean(vararg path: String): Boolean? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is Boolean) {
            rawData
        } else {
            throw ValueError("Data is not instance of Boolean")
        }
    }

    fun getObject(vararg path: String): Any? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return rawData ?: throw ValueError("Data is not instance of Object")
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getArrayList(vararg path: String): ArrayList<out T>? {
        val stringArray = ArrayList(listOf(*path))
        if (!saveData.containsKey(listOf(*path))) {
            return null
        }
        val rawData = saveData[stringArray]
        return if (rawData is ArrayList<*>?) {
            rawData as ArrayList<out T>?
        } else {
            throw ValueError("Data is not instance of ArrayList")
        }
    }
}